actioncable/lib/action_cable/connection/web_socket.rb

```diff
@@ -34,8 +32,21 @@ def rack_response
         websocket.rack_response
       end
 
+      # by order of preference, the first supported client will be chosen
+      CLIENT_SOCKET_CANDIDATES = [ClientRackSocket, ClientFayeSocket]
+
+      def self.establish_connection(env, event_target, event_loop, protocols)
+        @client_socket_klass ||= client_socket_selector(env)
+        @client_socket_klass&.attempt(env, event_target, event_loop, protocols)
+      end
```

Comment:
We already chose a client socket that accepts the connection, so we don't need the `attempt` class methods.

Reply:
`attempt` will test the headers to make sure this is a valid WebSocket response.

The headers are checked twice only when memoizing the `klass` type, afterwards we need attempt to make sure the WebSocket requests aren't malformed.

Reply:
> ... to make sure the WebSocket requests aren't malformed.

Why not just failing in that case with the exact failure reason? If we return `nil` we swallow the exception.

Reply:
@palkan , I don't think we should throw an exception.

The `attempt` method protects against a client error, not a server error. IMHO, the server should expect clients to be sometimes "misguided" (if not down-right malicious).

This way the connection is closed with the 400 error code, no exception is raised.

Reply:
Should I remove the memoizing logic? Or should we mark this change request as resolved?